src/sinks/util/buffer/loki.rs

```diff
@@ -0,0 +1,269 @@
+//! Special case Loki sink batch buffer
+//!
+//! This buffer handles stream merging -- when a record is inserted into
+//! the buffer, all records having the same stream label set are grouped
+//! together for more efficient output.
+
+use super::{
+    err_event_too_large, json::BoxedRawValue, Batch, BatchConfig, BatchError, BatchSettings,
+    BatchSize, PushResult,
+};
+use serde_json::{json, value::to_raw_value};
+use std::collections::HashMap;
+
+const WRAPPER_OVERHEAD: usize = r#"{"streams":[]}"#.len();
+const STREAM_OVERHEAD: usize = r#"{"stream":{},"values":[]}"#.len();
+const LABEL_OVERHEAD: usize = r#""":"""#.len();
+
+pub type Labels = Vec<(String, String)>;
+
+#[derive(Clone, Debug)]
+pub struct LokiEvent {
+    pub timestamp: i64,
+    pub event: String,
+}
+
+#[derive(Clone, Debug)]
+pub struct LokiRecord {
+    pub labels: Labels,
+    pub event: LokiEvent,
+}
+
+#[derive(Debug)]
+struct LokiEncodedEvent {
+    pub timestamp: i64,
+    pub encoded: BoxedRawValue,
+}
+
+impl From<&LokiEvent> for LokiEncodedEvent {
+    // Pre-encode the record to JSON, but keep the timestamp for sorting at the end.
+    // The final output should be: `[ts, line]'
+    fn from(event: &LokiEvent) -> Self {
+        Self {
+            timestamp: event.timestamp,
+            encoded: to_raw_value(&json!([format!("{}", event.timestamp), event.event]))
+                .expect("JSON encoding should never fail"),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct LokiBuffer {
+    num_bytes: usize,
+    num_items: usize,
+    streams: HashMap<Labels, Vec<LokiEncodedEvent>>,
+    settings: BatchSize<Self>,
+}
+
+impl LokiBuffer {
+    pub fn new(settings: BatchSize<Self>) -> Self {
+        Self {
+            num_bytes: WRAPPER_OVERHEAD,
+            num_items: 0,
+            streams: HashMap::default(),
+            settings,
+        }
+    }
+}
+
+impl Batch for LokiBuffer {
+    type Input = LokiRecord;
+    type Output = serde_json::Value;
+
+    fn get_settings_defaults(
+        config: BatchConfig,
+        defaults: BatchSettings<Self>,
+    ) -> Result<BatchSettings<Self>, BatchError> {
+        Ok(config
+            .use_size_as_events()?
+            .get_settings_or_default(defaults))
+    }
+
+    fn push(&mut self, mut item: Self::Input) -> PushResult<Self::Input> {
+        let labels_len = item
+            .labels
+            .iter()
+            .map(|label| label.0.len() + label.1.len() + LABEL_OVERHEAD)
+            .sum::<usize>()
```

Comment:
```suggestion
            .fold(0, |total, label| total + label.0.len() + label.1.len() + LABEL_OVERHEAD)
```

Reply:
I don't think that will either help code generation or expressiveness.